lib/miq_automation_engine/models/miq_ae_git.rb

```diff
@@ -0,0 +1,328 @@
+  end
+
+  def mv_file_with_new_contents(old_file, new_hash, commit_sha = nil)
+    add(new_hash, commit_sha)
+    remove(old_file, commit_sha)
+  end
+
+  def mv_file(old_file, new_file, commit_sha = nil)
+    entry = current_index[old_file]
+    return unless entry
+    entry[:path] = new_file
+    current_index(commit_sha).add(entry)
+    remove(old_file, commit_sha)
+  end
+
+  def mv_dir(old_dir, new_dir, commit_sha = nil)
+    old_dir = fix_path_mv(old_dir)
+    new_dir = fix_path_mv(new_dir)
+    updates = current_index(commit_sha).entries.select { |entry| entry[:path].start_with?(old_dir) }
+    updates.each do |entry|
+      entry[:path] = entry[:path].gsub(old_dir, new_dir)
+      current_index(commit_sha).add(entry)
+    end
+    current_index(commit_sha).remove_dir(old_dir)
+  end
+
+  private
+
+  def fix_path_mv(dir_name)
+    dir_name = dir_name[1..-1] if dir_name[0] == '/'
+    dir_name += '/'            if dir_name[-1] != '/'
+    dir_name
+  end
+
+  def filename(path, commit_sha = nil)
+    entry = get_tree_entry(path, commit_sha)
+    entry ? entry[:full_name] : nil
+  end
+
+  def get_tree(path, commit_sha = nil)
+    entry = get_tree_entry(path, commit_sha)
+    raise MiqException::MiqGitEntryMissing, path unless entry
+    raise MiqException::MiqGitEntryNotADirectory, path  unless entry[:type] == :tree
+    @repo.lookup(entry[:oid])
+  end
+
+  def lookup_commit_tree(commit_sha = nil)
+    return nil unless @repo.branches['master']
+    ct = commit_sha ? @repo.lookup(commit_sha) : @repo.branches['master'].target
+    ct.tree if ct
+  end
+
+  def get_tree_entry(path, commit_sha = nil)
+    path = path[1..-1] if path[0] == '/'
+    tree = lookup_commit_tree(commit_sha || @commit_sha)
+    begin
+      entry             = tree.path(path)
+      entry[:full_name] = File.join(@base_name, path)
+      entry[:rel_path]  = path
+    rescue
+      return nil
+    end
+    entry
+  end
+
+  def current_index(commit_sha = nil)
+    @current_index ||= Rugged::Index.new.tap do |index|
+      unless @repo.empty?
+        tree = lookup_commit_tree(commit_sha || @commit_sha)
+        raise ArgumentError, "Cannot locate commit tree" unless tree
+        @current_tree_oid = tree.oid
+        index.read_tree(tree)
+      end
+    end
+  end
+
+  def create_commit(message, tree, parents)
+    author = {:email => @email, :name => @name, :time => Time.now}
+    # Create the actual commit but dont update the reference
+    Rugged::Commit.create(@repo, :author  => author,  :committer  => author,
+                                 :message => message, :parents    => parents,
+                                 :tree    => tree)
+  end
+
+  def lock
+    @repo.references.create(LOCK_REFERENCE, 'refs/heads/master')
+    yield
+    rescue Rugged::ReferenceError
+      sleep 0.1
+      retry
+    ensure
+      @repo.references.delete(LOCK_REFERENCE)
+  end
+
+  def push_lock
+    @repo.references.create(LOCK_REFERENCE, 'refs/heads/master')
```

Comment:
This pattern of lock/yield/remove_lock should move into a method like `with_lock`